presto-main/src/test/java/com/facebook/presto/operator/TestOperationTimer.java

```diff
@@ -0,0 +1,177 @@
+        long previousOperation2WallTime = operationTiming2.getWallNanos();
+
+        doSomething();
+        timer.recordOperationComplete(operationTiming1);
+        assertEquals(operationTiming1.getCalls(), 3);
+        assertThat(operationTiming1.getCpuNanos()).isGreaterThan(previousOperation1CpuTime);
+        assertThat(operationTiming1.getWallNanos()).isGreaterThan(previousOperation1WallTime);
+
+        doSomething();
+        timer.recordOperationComplete(operationTiming2);
+        assertEquals(operationTiming2.getCalls(), 2);
+        assertThat(operationTiming2.getCpuNanos()).isGreaterThan(previousOperation2CpuTime);
+        assertThat(operationTiming2.getWallNanos()).isGreaterThan(previousOperation2WallTime);
+
+        doSomething();
+        timer.recordOperationComplete(operationTiming3);
+        assertEquals(operationTiming3.getCalls(), 1);
+        assertThat(operationTiming3.getCpuNanos()).isGreaterThan(0);
+        assertThat(operationTiming3.getWallNanos()).isGreaterThan(operationTiming1.getCpuNanos());
+
+        timer.end(overallTiming);
+
+        assertThat(operationTiming1.getWallNanos() + operationTiming2.getWallNanos() + operationTiming3.getWallNanos()).isLessThanOrEqualTo(overallTiming.getWallNanos());
+        assertThat(operationTiming1.getCpuNanos() + operationTiming2.getCpuNanos() + operationTiming3.getCpuNanos()).isLessThanOrEqualTo(overallTiming.getCpuNanos());
+    }
+
+    @Test
+    public void testOperationCpuTrackingDisabled()
+    {
+        OperationTiming overallTiming = new OperationTiming();
+        OperationTiming operationTiming = new OperationTiming();
+
+        OperationTimer timer = new OperationTimer(true, false);
+
+        doSomething();
+        timer.recordOperationComplete(operationTiming);
+        assertEquals(operationTiming.getCalls(), 1);
+        assertEquals(operationTiming.getCpuNanos(), 0);
+        assertThat(operationTiming.getWallNanos()).isGreaterThan(0);
+
+        timer.end(overallTiming);
+
+        assertEquals(overallTiming.getCalls(), 1);
+        assertThat(overallTiming.getCpuNanos()).isGreaterThan(0);
+        assertThat(overallTiming.getWallNanos()).isGreaterThan(overallTiming.getCpuNanos());
+    }
+
+    @Test(expectedExceptions = IllegalStateException.class)
+    public void testOverallTrackingTwice()
+    {
+        OperationTiming timing = new OperationTiming();
+        OperationTimer timer = new OperationTimer(true, false);
+        timer.end(timing);
+        timer.end(timing);
+    }
+
+    private static void doSomething()
+    {
+        spin();
+        sleep();
+    }
+
+    private static void sleep()
+    {
+        try {
+            Thread.sleep(50);
+        }
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static void spin()
+    {
+        for (int i = 0; i < 1_000_000; i++) {
```

Comment:
This is not a reliable way to do busy work. The VM could well fold this whole thing into a constant-time operation.

Reply:
Based on my anecdotal observation it doesn't. But due to the memory model the JVM can technically do that.

Reply:
I didn't realize i hadn't made the variable to be `volatile`. If it is volatile, java cannot fold it. To increment the volatile value it should read the value every time.

Reply:
Try `xxhash` (or something even slower) on a chunk of data and store the result in a public volatile?

Reply:
Instead of relying on fragile assumptions, how about we modify `OperationTimer` to use a test thread cpu time provider (instead of always using `THREAD_MX_BEAN`)? We can then simply increment some thread cpu time counters in these tests and provide those values to the `OperationTimer`.

Reply:
> Instead of relying on fragile assumptions, how about we modify OperationTimer to use a test thread cpu time provider (instead of always using THREAD_MX_BEAN)? We can then simply increment some thread cpu time counters in these tests and provide those values to the OperationTimer.

That's a good idea. I just didn't want to complicate the `OperationTimer` code to simplify testing. I changed the `spin()` code to taking a hash from randomly generated 10kb of data using a volatile random seed and storing it back to the volatile variable.